File: scala/PeppolLookup.scala
/**
 * PEPPOL uses two key services to enable document exchange:
 *
 * 1. SML (Service Metadata Locator):
 *    - Acts as a DNS-based directory service
 *    - Maps a participant's ID to their SMP provider
 *    - Uses DNS lookup to find where a participant's metadata is hosted
 *    - Similar to how email's MX records help find mail servers
 *
 * 2. SMP (Service Metadata Publisher):
 *    - Hosts metadata about a participant's capabilities
 *    - Tells you what document types they can receive
 *    - Provides technical details needed for sending documents
 *    - Acts like a participant's business card in the network
 *
 * This example demonstrates how to:
 * 1. Use SML to find where a participant's metadata is hosted
 * 2. Query their SMP to discover what documents they can receive
 * 3. Check for PEPPOL BIS Billing 3.0 support
 */

import java.net.{InetAddress, URL, URLEncoder, URLDecoder, UnknownHostException}
import java.security.MessageDigest
import scala.io.Source
import scala.util.{Try, Using}

object PeppolLookup {
  // Test environment SML domain
  val SML_DOMAIN = "edelivery.tech.ec.europa.eu"

  // PEPPOL BIS Billing 3.0 document identifiers
  val BIS_BILLING_INVOICE = "urn:oasis:names:specification:ubl:schema:xsd:Invoice-2::Invoice"
  val BIS_BILLING_CREDITNOTE = "urn:oasis:names:specification:ubl:schema:xsd:CreditNote-2::CreditNote"

  /**
   * Step 1: Use SML (Service Metadata Locator) to find a participant's SMP hostname
   *
   * The SML is like a phone book for the PEPPOL network. Given a participant's ID:
   * 1. Create an MD5 hash of their ID (e.g., "0192:921605900")
   * 2. Use the hash to construct a DNS hostname
   * 3. If the hostname exists, the participant is registered in PEPPOL
   * 4. The hostname tells us where to find their metadata (SMP)
   *
   * Returns Some(hostname) if found, None if not found
   */
  def smlLookup(icd: String, identifier: String, smlDomain: String = SML_DOMAIN): Option[String] = {
    Try {
      // Create MD5 hash of participant ID
      val participantId = s"$icd:$identifier"
      val md5 = MessageDigest.getInstance("MD5")
      val hashBytes = md5.digest(participantId.getBytes)
      val md5Hash = hashBytes.map("%02x".format(_)).mkString

      // Construct hostname
      val hostname = s"b-$md5Hash.iso6523-actorid-upis.$smlDomain"

      // Check if hostname exists
      InetAddress.getByName(hostname)
      hostname
    }.toOption
  }

  /**
   * Step 2: Query SMP (Service Metadata Publisher) to get supported document types
   *
   * The SMP is like a business card in the PEPPOL network. It tells us:
   * 1. What types of documents the participant can receive
   * 2. Technical details needed for sending documents
   * 3. Specific document format versions they support
   *
   * This is similar to how DNS MX records tell you where to send email,
   * but SMP also includes what "types" of messages you can send.
   */
  def smpLookup(smpHostname: String, icd: String, identifier: String): List[String] = {
    // Construct SMP URL
    // Format: http://[SMP hostname]/[identifier scheme]::[participant identifier]
    val participantId = s"$icd:$identifier"
    val encodedId = URLEncoder.encode(participantId, "UTF-8")
    val urlString = s"http://$smpHostname/iso6523-actorid-upis::$encodedId"

    // Perform HTTP GET request
    val url = new URL(urlString)
    val response = Using(Source.fromURL(url)) { source =>
      source.mkString
    }.get

    // Extract document types from ServiceMetadataReference href attributes
    val pattern = """ServiceMetadataReference[^>]*href="([^"]*)"[^>]*>""".r

    pattern.findAllMatchIn(response).flatMap { m =>
      val href = URLDecoder.decode(m.group(1), "UTF-8")
      if (href.contains("busdox-docid-qns::")) {
        val docType = href.split("busdox-docid-qns::")(1).split("#")(0)
        Some(docType)
      } else {
        None
      }
    }.toList
  }

  def main(args: Array[String]): Unit = {
    try {
      // Snapbooks AS (Norwegian organization number)
      val icd = "0192"
      val identifier = "921605900"

      // Step 1: Use SML to find where participant's metadata is hosted
      smlLookup(icd, identifier) match {
        case None =>
          println(s"Not a PEPPOL participant: $icd:$identifier")
          sys.exit(1)

        case Some(smpHostname) =>
          println(s"SMP hostname: $smpHostname")

          // Step 2: Query their SMP to discover supported documents
          val documentTypes = smpLookup(smpHostname, icd, identifier)
          println("\nSupported document identifiers:")
          documentTypes.foreach(docType => println(s"- $docType"))

          // Check for PEPPOL BIS Billing 3.0 documents
          println("\nPEPPOL BIS Billing 3.0 Support:")
          if (documentTypes.contains(BIS_BILLING_INVOICE)) {
            println("- Supports Invoice")
          }
          if (documentTypes.contains(BIS_BILLING_CREDITNOTE)) {
            println("- Supports Credit Note")
          }
      }
    } catch {
      case e: Exception =>
        System.err.println(s"Error: ${e.getMessage}")
        sys.exit(1)
    }
  }
}

File: clojure/peppol_lookup.clj
#!/usr/bin/env clojure

;; PEPPOL uses two key services to enable document exchange:
;;
;; 1. SML (Service Metadata Locator):
;;    - Acts as a DNS-based directory service
;;    - Maps a participant's ID to their SMP provider
;;    - Uses DNS lookup to find where a participant's metadata is hosted
;;    - Similar to how email's MX records help find mail servers
;;
;; 2. SMP (Service Metadata Publisher):
;;    - Hosts metadata about a participant's capabilities
;;    - Tells you what document types they can receive
;;    - Provides technical details needed for sending documents
;;    - Acts like a participant's business card in the network
;;
;; This example demonstrates how to:
;; 1. Use SML to find where a participant's metadata is hosted
;; 2. Query their SMP to discover what documents they can receive
;; 3. Check for PEPPOL BIS Billing 3.0 support

(import '(java.net InetAddress UnknownHostException URL URLEncoder URLDecoder)
        '(java.security MessageDigest))

;; Test environment SML domain
(def sml-domain "edelivery.tech.ec.europa.eu")

;; PEPPOL BIS Billing 3.0 document identifiers
(def bis-billing-invoice "urn:oasis:names:specification:ubl:schema:xsd:Invoice-2::Invoice")
(def bis-billing-creditnote "urn:oasis:names:specification:ubl:schema:xsd:CreditNote-2::CreditNote")

;; Helper function to convert bytes to hex string
(defn bytes-to-hex [bytes]
  (apply str (map #(format "%02x" %) bytes)))

;; Step 1: Use SML (Service Metadata Locator) to find a participant's SMP hostname
;;
;; The SML is like a phone book for the PEPPOL network. Given a participant's ID:
;; 1. Create an MD5 hash of their ID (e.g., "0192:921605900")
;; 2. Use the hash to construct a DNS hostname
;; 3. If the hostname exists, the participant is registered in PEPPOL
;; 4. The hostname tells us where to find their metadata (SMP)
;;
;; Returns the SMP hostname if found, nil if not found
(defn sml-lookup [icd identifier]
  (try
    ;; Create MD5 hash of participant ID
    (let [participant-id (str icd ":" identifier)
          md5 (MessageDigest/getInstance "MD5")
          hash-bytes (.digest md5 (.getBytes participant-id))
          md5-hash (bytes-to-hex hash-bytes)

          ;; Construct hostname
          hostname (str "b-" md5-hash ".iso6523-actorid-upis." sml-domain)]

      ;; Check if hostname exists
      (InetAddress/getByName hostname)
      hostname)
    (catch UnknownHostException _ nil)))

;; Step 2: Query SMP (Service Metadata Publisher) to get supported document types
;;
;; The SMP is like a business card in the PEPPOL network. It tells us:
;; 1. What types of documents the participant can receive
;; 2. Technical details needed for sending documents
;; 3. Specific document format versions they support
;;
;; This is similar to how DNS MX records tell you where to send email,
;; but SMP also includes what "types" of messages you can send.
(defn smp-lookup [smp-hostname icd identifier]
  ;; Construct SMP URL
  ;; Format: http://[SMP hostname]/[identifier scheme]::[participant identifier]
  (let [participant-id (str icd ":" identifier)
        encoded-id (URLEncoder/encode participant-id "UTF-8")
        url-str (str "http://" smp-hostname "/iso6523-actorid-upis::" encoded-id)
        url (URL. url-str)]

    ;; Perform HTTP GET request
    (with-open [stream (.openStream url)]
      (let [response (slurp stream)

            ;; Extract document types from ServiceMetadataReference href attributes
            pattern #"ServiceMetadataReference[^>]*href=\"([^\"]*)\"[^>]*>"
            matches (re-seq pattern response)]

        (->> matches
             (map second)
             (map #(URLDecoder/decode % "UTF-8"))
             (filter #(re-find #"busdox-docid-qns::" %))
             (map #(second (clojure.string/split % #"busdox-docid-qns::")))
             (map #(first (clojure.string/split % #"#")))
             (vec))))))

;; Main execution
(defn -main []
  (try
    ;; Snapbooks AS (Norwegian organization number)
    (let [icd "0192"
          identifier "921605900"]

      ;; Step 1: Use SML to find where participant's metadata is hosted
      (if-let [smp-hostname (sml-lookup icd identifier)]
        (do
          (println (str "SMP hostname: " smp-hostname))

          ;; Step 2: Query their SMP to discover supported documents
          (let [document-types (smp-lookup smp-hostname icd identifier)]
            (println "\nSupported document identifiers:")
            (doseq [doc-type document-types]
              (println (str "- " doc-type)))

            ;; Check for PEPPOL BIS Billing 3.0 documents
            (println "\nPEPPOL BIS Billing 3.0 Support:")
            (when (some #{bis-billing-invoice} document-types)
              (println "- Supports Invoice"))
            (when (some #{bis-billing-creditnote} document-types)
              (println "- Supports Credit Note"))))

        (do
          (println (str "Not a PEPPOL participant: " icd ":" identifier))
          (System/exit 1))))

    (catch Exception e
      (binding [*out* *err*]
        (println (str "Error: " (.getMessage e))))
      (System/exit 1))))

;; Run main
(-main)
